clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -3675,14 +4062,260 @@ class SyclKernelBodyCreator : public SyclKernelFieldHandler {
     removeFieldMemberExpr(FD, ArrayType);
     return true;
   }
+  using SyclKernelFieldHandler::enterArray;
+  using SyclKernelFieldHandler::enterStruct;
+  using SyclKernelFieldHandler::handleNonDecompStruct;
+  using SyclKernelFieldHandler::handlePointerType;
+  using SyclKernelFieldHandler::handleScalarType;
+  using SyclKernelFieldHandler::handleSyclSpecialType;
+  using SyclKernelFieldHandler::handleUnionType;
+  using SyclKernelFieldHandler::leaveArray;
+  using SyclKernelFieldHandler::leaveStruct;
+};
+
+class FreeFunctionKernelBodyCreator : public SyclKernelFieldHandler {
+  SyclKernelDeclCreator &DeclCreator;
+  llvm::SmallVector<Stmt *, 16> BodyStmts;
+  FunctionDecl *FreeFunc;
+  SourceLocation KernelCallerSrcLoc; // KernelCallerFunc source location.
+  llvm::SmallVector<Expr *, 8> ArgExprs;
+
+  // Creates a DeclRefExpr to the ParmVar that represents the current field.
+  Expr *createParamReferenceExpr() {
+    ParmVarDecl *KernelParameter =
+        DeclCreator.getParamVarDeclsForCurrentField()[0];
+
+    QualType ParamType = KernelParameter->getOriginalType();
+    Expr *DRE = SemaSYCLRef.SemaRef.BuildDeclRefExpr(
+        KernelParameter, ParamType, VK_PRValue, KernelCallerSrcLoc);
+    return DRE;
+  }
+
+  // Creates a DeclRefExpr to the ParmVar that represents the current pointer
+  // field.
+  Expr *createPointerParamReferenceExpr(QualType PointerTy, bool Wrapped) {
+    ParmVarDecl *KernelParameter =
+        DeclCreator.getParamVarDeclsForCurrentField()[0];
+
+    QualType ParamType = KernelParameter->getOriginalType();
+    Expr *DRE = SemaSYCLRef.SemaRef.BuildDeclRefExpr(
+        KernelParameter, ParamType, VK_LValue, KernelCallerSrcLoc);
+    DRE = ImplicitCastExpr::Create(SemaSYCLRef.getASTContext(), ParamType,
+                                   CK_LValueToRValue, DRE, /*BasePath=*/nullptr,
+                                   VK_PRValue, FPOptionsOverride());
+
+    if (PointerTy->getPointeeType().getAddressSpace() !=
+        ParamType->getPointeeType().getAddressSpace())
+      DRE = ImplicitCastExpr::Create(SemaSYCLRef.getASTContext(), PointerTy,
+                                     CK_AddressSpaceConversion, DRE, nullptr,
+                                     VK_PRValue, FPOptionsOverride());
+
+    return DRE;
+  }
+
+  // For a free function such as:
+  // void f(int i, int* p, struct Simple S) { … }
+  //
+  // Keep the function as-is for the version callable from device code.
+  // void f(int i, int *p, struct Simple S) { … }
+  //
+  // For the host-callable kernel function generate this:
+  // void __free_function_f(int _arg_i, int* _arg_p, struct Simple _arg_S)
+  // {
+  //   f(_arg_i, _arg_p, _arg_S);
+  // }
+  CompoundStmt *createFreeFunctionKernelBody() {
+    SemaSYCLRef.SemaRef.PushFunctionScope();
+    Expr *Fn = SemaSYCLRef.SemaRef.BuildDeclRefExpr(
+        FreeFunc, FreeFunc->getType(), VK_PRValue, KernelCallerSrcLoc);
```

Comment:
Why `VK_PRValue` here?

Reply:
VK_LValue didn't work. We're just reading the address, right? Not assigning something to it.